src/Compilers/CSharp/Portable/Lowering/InstanceExtensionMethodReferenceRewriter.cs

```diff
@@ -0,0 +1,219 @@
+            _containingMethod = node.Symbol;
+
+            var result = base.VisitLocalFunctionStatement(node);
+
+            _containingMethod = oldContainingMethod;
+            return result;
+        }
+
+        public override BoundNode VisitCall(BoundCall node)
+        {
+            Debug.Assert(node != null);
+
+            BoundExpression rewrittenCall;
+
+            if (LocalRewriter.TryGetReceiver(node, out BoundCall? receiver1))
+            {
+                // Handle long call chain of both instance and extension method invocations.
+                var calls = ArrayBuilder<BoundCall>.GetInstance();
+
+                calls.Push(node);
+                node = receiver1;
+
+                while (LocalRewriter.TryGetReceiver(node, out BoundCall? receiver2))
+                {
+                    calls.Push(node);
+                    node = receiver2;
+                }
+
+                // Rewrite the receiver
+                BoundExpression? rewrittenReceiver = (BoundExpression?)this.Visit(node.ReceiverOpt);
+
+                do
+                {
+                    rewrittenCall = visitArgumentsAndFinishRewrite(node, rewrittenReceiver);
+                    rewrittenReceiver = rewrittenCall;
+                }
+                while (calls.TryPop(out node!));
+
+                calls.Free();
+            }
+            else
+            {
+                // Rewrite the receiver
+                BoundExpression? rewrittenReceiver = (BoundExpression?)this.Visit(node.ReceiverOpt);
+                rewrittenCall = visitArgumentsAndFinishRewrite(node, rewrittenReceiver);
+            }
+
+            return rewrittenCall;
+
+            BoundExpression visitArgumentsAndFinishRewrite(BoundCall node, BoundExpression? rewrittenReceiver)
+            {
+                return UpdateCall(
+                    _containingMethod,
+                    node,
+                    this.VisitMethodSymbol(node.Method),
+                    this.VisitSymbols<MethodSymbol>(node.OriginalMethodsOpt),
+                    rewrittenReceiver,
+                    this.VisitList(node.Arguments),
+                    node.ArgumentRefKindsOpt,
+                    node.InvokedAsExtensionMethod,
+                    this.VisitType(node.Type));
+            }
+        }
+
+        internal static BoundExpression UpdateCall(
+            MethodSymbol containingMethod,
+            BoundCall boundCall,
+            MethodSymbol method,
+            ImmutableArray<MethodSymbol> originalMethodsOpt,
+            BoundExpression? receiverOpt,
+            ImmutableArray<BoundExpression> arguments,
+            ImmutableArray<RefKind> argumentRefKinds,
+            bool invokedAsExtensionMethod,
+            TypeSymbol type)
+        {
+            var temps = ImmutableArray<LocalSymbol>.Empty;
+
+            if (receiverOpt is not null && method.OriginalDefinition.ContainingSymbol is NamedTypeSymbol declaringTypeDefinition &&
```

Comment:
Consider inverting the condition and returning early (without update) when possible.

Reply:
> Consider inverting the condition and returning early (without update) when possible.

It is quite possible I misunderstand the suggestion, but we cannot skip update, ever, because the arguments and the receiver might not match what we have in the original BoundCall, they have been rewritten. In this `if` we might do additional adjustments. Given that, I do not see much advantage in inverting the condition.